restli-client/src/main/java/com/linkedin/restli/client/ExecutionGroup.java

```diff
@@ -0,0 +1,167 @@
+ *       {@code <YourClient>}.get({@code <Parameter2>}); // implicitly add to the ExecutionGroup which created this lambda
+ *       new ExecutionGroup().batchOn(() -> {
+ *         {@code <YourClient2>}.get({@code <Parameter3>});
+ *         {@code <YourClient2>}.get({@code <Parameter4>});
+ *       }); // this execution group will not be affecting the outer execution group.
+ *       // adding to eg3 explicitly so will not be batched together with other implicit calls in this lambda clause.
+ *       {@code <YourClient>}.get({@code <Parameter3>}, eg3);
+ *     }); // get call from {@code <YourClient>} with parameter1 and parameter2 will be batched
+ *   </pre>
+ * </blockquote>
+ *
+ *
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ExecutionGroup
+{
+  private final Map<FluentClient, List<Task>> _clientToTaskListMap = new HashMap<>();
+  private final Engine _engine;
+
+  private Set<FluentClient> _fluentClientAll; // filled by UClient when executionGroup is created; Used for batchOn
+
+  /**
+   * This constructor will be called by the UniversalClient and will not be called by API users directly
+   * @param engine
+   */
+  /* package private */ ExecutionGroup(Engine engine)
+  {
+    _engine = engine;
+  }
+
+  /**
+   * Execute all the tasks that added to {@link ExecutionGroup} through ParSeq Engine
+   */
+  public void execute()
+  {
+    for (Map.Entry<FluentClient, List<Task>> entry : _clientToTaskListMap.entrySet()) {
+      List<Task> taskList = entry.getValue();
+      // the Task.par(Iterable) version does not fast-fail comparing to Task.par(Task...)
+      ParTask<Object> perFluentClientTasks =
+          Task.par(taskList.stream().map(obj -> ((Task<Object>) obj)).collect(Collectors.toList()));
+      taskList.clear();
+      _clientToTaskListMap.remove(entry.getKey());
+      // starts a plan for tasks from one client due to performance consideration
+      // TODO: optimize, use scheduleAndRun
+      _engine.run(perFluentClientTasks);
+    }
+  }
+
+  /**
+   * Run user's logic provided in lambda function and batch related requests made using FluentAPI inside this lambda function
+   *
+   * Note the FluentAPI requests that take the ExecutionGroup instance as an explicit parameter will not be batched.
+   * Also everytime this method is called, it creates a separate ExecutionGroup so nested ExecutionGroup won't be affected.
+   *
+   * @param runnable the runnable that executes user's logic
+   * @param fluentClients the fluentClients whose requests will be batched, if None specified, all fluentClients call
+   *                      will be batched.
+   * @throws Exception
+   */
+  public void batchOn(Runnable runnable, FluentClient... fluentClients) throws Exception
+  {
+    List<FluentClient> batchedClients =
+        fluentClients.length > 0 ? new ArrayList<FluentClient>(Arrays.asList(fluentClients))
+            : new ArrayList<FluentClient>(_fluentClientAll);
```

Comment:
Store _fluentClientAll as a List to avoid creating this everytime.

Reply:
taken. Updated